Log elasticsearch issues in any case

JoshuaBehrens opened the pull request:
### 1. Why is this change necessary?
When you run in prod mode and have an elastichsearch falling back to database search you don't know why because neither a log message is logged or an exception is thrown.

### 2. What does this change do, exactly?
Log in any case of an exception before environment evaluation is done.

### 3. Describe each step to reproduce the issue or behaviour.
1. Use plugins that extend elasticsearch queries
2. Have a slow page or weird behaviour when enabling elastic based search
3. No hints in the log files

### 4. Checklist
Is there a legit way how to test it when the first check literally is: am I in a test?

- [ ] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-Implement-New-Changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

JoshuaBehrens committed:
Log elasticsearch issues in any case

JoshuaBehrens commented:
I'll add a changelog file later

JoshuaBehrens commented:
> Is there a legit way how to test it when the first check literally is: am I in a test?

For real? Shouldn't the first be a mock or decorator doing exactly this?

shyim commented:
@JoshuaBehrens You mean with testing? You can construct the class by own and pass a mock to test it

JoshuaBehrens commented:
@shyim I mean why is there an env check on `'test'`? This should not be code within this namespace, right?

NiklasLimberg commented:
I'll mark this as incomplete as long as there isn't a changelog file

NiklasLimberg commented:
Hi @JoshuaBehrens,
please write a quick changelog for this and make sure to include the ticket number

JoshuaBehrens commented:
I added a changelog. Reading that changelog makes me think: do we want to rename `logOrThrowException` to `tryToThrow`? I mean logging is not an optional case anymore

mitelg commented on src/Elasticsearch/Framework/ElasticsearchHelper.php:
```diff
@@ -63,6 +63,8 @@ public function __construct(
 
     public function logOrThrowException(\Throwable $exception): bool
```
now the method name does not say anymore what really happens in here :grin:

JoshuaBehrens replied:
@mitelg that was my previous comment about https://github.com/shopware/platform/pull/2035#issuecomment-920739272 so shall we change it as well?

mitelg replied:
oh, I did not read that :see_no_evil: 

renaming the method would be a breaking change. what is possible, is to add a new method and deprecate this one which then only calls the new one

JoshuaBehrens replied:
Yes sure. I would follow the deprecation guide :) I can do that

JoshuaBehrens replied:
I hate myself already for this as I decorated it and it probably would break sth in my decoration... :(

mitelg replied:
maybe we need to introduce an abstract class for that? :thinking: @keulinho @shyim @OliverSkroblin what do you think?

NiklasLimberg replied:
Ok you'll figure this one out @mitelg @JoshuaBehrens. Just hit me up when this is ready to be imported

JoshuaBehrens replied:
@mitelg do you think about to create a Decoratable?

shyim replied:
I am fine with an abstract class.

JoshuaBehrens replied:
So just abstract class? What is not abstract on it? When it is an abstract class, shall I make it a decoratable? Otherwise an interface would be sufficient right?

shyim replied:
we don't make interfaces anymore and use abstract classes for decoration pattern. see https://github.com/shopware/platform/blob/trunk/src/Core/Content/Seo/AbstractSeoResolver.php#L5 as example

JoshuaBehrens replied:
That was what I tried to find out. I make it a decoratable then

NiklasLimberg commented:
well this was already imported to jira hmmmm

JoshuaBehrens commented:
@NiklasLimberg which one shall I reference in the changelog?

NiklasLimberg commented:
@JoshuaBehrens Ideally NEXT-16905, i've closed the otherone as a duplicate already

JoshuaBehrens commented:
I rebased, refactored the helper into an abstract one, aliased the service id, changed changelog. Can you tell me that it was ok to not pull setEnabled into the contract? This isn't needed anyway when https://github.com/shopware/platform/issues/1821 is resolved, right?

@mitelg did I do that correctly?

NiklasLimberg commented:
Just disregard the new issue, I forgot again that it already has an issue

shyim commented on src/Elasticsearch/Framework/AbstractElasticsearchHelper.php:
```diff
@@ -0,0 +1,52 @@
+abstract class AbstractElasticsearchHelper
+{
+    /**
+     * @throws DecorationPatternException
```
I would not mention the throws in the abstract class. as this should do only the core implementation.

JoshuaBehrens replied:
I rather would. Anyone who decorates it needs to expect that to happen. It is basically part of any decoratable contract. But you are member, so I can remove it.

shyim commented:
As we discussed on Slack, remove the abstract class :+1:

JoshuaBehrens commented:
I rebased and amend by consensus

NiklasLimberg commented:
Hi @JoshuaBehrens,
i've imported this aswell again @shyim will review it once he is back from vacation next week

shyim commented:
Thanks for your contribution
